# 📊 E-commerce Returns & Profitability Analysis

End-to-end analysis of Brazilian e-commerce dataset to uncover insights on returns, profitability, and operational efficiency.


## 🧹 Data Cleaning


### 📌 Environment Setup

This project requires the following Python libraries:
- SQLAlchemy for database connection
- Psycopg2 for PostgreSQL integration
- Pandas for data manipulation

All dependencies are listed in the `requirements.txt` file.

In [1]:
!pip install sqlalchemy psycopg2

## 🔍 Exploratory Data Analysis (EDA)


### 📌 Database Connection Validation

In this step, we validate the connection between Python and PostgreSQL using a simple test query.

The query `SELECT 1` is commonly used to verify that:
- The database server is running
- Credentials are correct
- The connection engine is properly configured

In [2]:
from sqlalchemy import create_engine
import pandas as pd

engine = create_engine(
    "postgresql://postgres:test123@localhost:5432/olist_project"
)

df = pd.read_sql("SELECT 1 as test", engine)
df

,test
0,1


### 📂 Setting the Working Directory and Verifying Dataset Files

In this step, we set the working directory to the location where all the Olist dataset files are stored. This ensures that Python can easily access the data files without needing full file paths every time.

We then list all the files present in the directory to confirm that the required datasets are available and correctly placed. This is an important validation step before proceeding with data loading and analysis.

In [5]:
import os

os.chdir(r"C:\Users\Anirban Tarafdar\Documents\Brazilian E-Commerce Public Dataset (Olist)")

os.listdir()

['master_olist.csv',
 'olist_customers_dataset.csv',
 'olist_geolocation_dataset.csv',
 'olist_orders_dataset.csv',
 'olist_order_items_dataset.csv',
 'olist_order_payments_dataset.csv',
 'olist_order_reviews_dataset.csv',
 'olist_products_dataset.csv',
 'olist_sellers_dataset.csv',
 'product_category_name_translation.csv']

### ✅ Output Interpretation

The output displays all files available in the specified directory. We can confirm that all required Olist dataset files, including orders, customers, products, payments, and reviews, are present.

This confirms that the environment is correctly set up, and we are ready to proceed with loading and analyzing the datasets.

### 📥 Loading Olist Dataset Files into Pandas DataFrames

In this step, we load all the relevant Olist dataset files into separate Pandas DataFrames. Each dataset represents a different aspect of the e-commerce ecosystem, such as customers, orders, products, payments, and reviews.

Loading the data into structured DataFrames allows us to efficiently perform data cleaning, transformation, and analysis in subsequent steps. Each DataFrame will later be used for relational analysis and SQL-based operations.

In [6]:
import pandas as pd

customers = pd.read_csv("olist_customers_dataset.csv")
orders = pd.read_csv("olist_orders_dataset.csv")
items = pd.read_csv("olist_order_items_dataset.csv")
payments = pd.read_csv("olist_order_payments_dataset.csv")
reviews = pd.read_csv("olist_order_reviews_dataset.csv")
products = pd.read_csv("olist_products_dataset.csv")
translation = pd.read_csv("product_category_name_translation.csv")

### ✅ Output Interpretation

The datasets have been successfully loaded into their respective DataFrames. Each variable now holds structured tabular data corresponding to a specific domain of the Olist e-commerce system.

These DataFrames will serve as the foundation for further data preprocessing, validation, and integration across multiple tables to derive meaningful business insights.

### 📊 Checking Dataset Dimensions

In this step, we examine the shape of the loaded datasets to understand their size in terms of rows and columns. This helps in assessing the volume of data and the number of features available in each dataset.

Understanding dataset dimensions is essential for planning further data processing, memory management, and analysis strategies.

In [7]:
customers.shape
orders.shape

(99441, 8)

### ✅ Output Interpretation

The output shows the dimensions of the datasets in the format (rows, columns).

- The customers dataset contains 99,441 records with 5 features.
- The orders dataset contains 99,441 records with 8 features.

This indicates a one-to-one relationship between customers and orders at the dataset level, which will be useful for joining and relational analysis in later stages.

### 🧾 Inspecting Column Names of the Datasets

In this step, we examine the column names of the datasets to understand the structure and available features. This helps in identifying key variables, primary keys, and potential relationships between different datasets.

Reviewing column names is an essential step before performing data cleaning, transformations, and joins across multiple tables.

In [8]:
customers.columns
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='object')

### ✅ Output Interpretation

The output displays the column names of the datasets.

For the orders dataset, key columns include:
- **order_id**: Unique identifier for each order
- **customer_id**: Links orders to customers
- **order_status**: Current status of the order (e.g., delivered, shipped)
- **order_purchase_timestamp**: Time when the order was placed
- **order_approved_at**: Time when the payment was approved
- **order_delivered_carrier_date**: Time when the order was handed to the carrier
- **order_delivered_customer_date**: Time when the order was delivered to the customer
- **order_estimated_delivery_date**: Expected delivery date

These columns are crucial for analyzing order lifecycle, delivery performance, and customer behavior in later stages.

### 🔍 Revisiting Customers Dataset Structure

In this step, we re-check the column names and dimensions of the customers dataset. This helps reinforce our understanding of its structure and ensures consistency before moving forward with data analysis.

Such validation steps are useful when working with multiple datasets to avoid confusion during joins and transformations.

In [9]:
customers.columns
customers.shape

(99441, 5)

### ✅ Output Interpretation

The customers dataset contains 99,441 rows and 5 columns.

This indicates that each record represents a unique customer entry with a limited set of attributes. The dataset is relatively compact and will primarily be used to link customer information with orders for relational analysis.

### 🧾 Exploring Customers Dataset Columns

In this step, we examine the column names of the customers dataset to understand the type of information available for each customer.

This helps in identifying key identifiers and location-based attributes, which are essential for customer segmentation and geographic analysis.

In [10]:
customers.columns

Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='object')

In [ ]:
### ✅ Output Interpretation

The customers dataset consists of the following columns:

- **customer_id**: Unique identifier for each customer in the dataset
- **customer_unique_id**: Identifier representing a unique individual (useful for identifying repeat customers)
- **customer_zip_code_prefix**: Prefix of the customer's zip code
- **customer_city**: City where the customer is located
- **customer_state**: State where the customer resides

These fields are particularly useful for analyzing customer distribution, regional trends, and repeat purchase behavior in later stages of the project.

### 🕒 Converting Date Columns to Datetime Format

In this step, we convert relevant columns in the orders dataset to datetime format. These columns represent different stages in the order lifecycle, such as purchase, approval, shipping, and delivery.

Converting these fields to datetime ensures accurate time-based analysis, enabling us to calculate metrics like delivery duration, processing time, and delays.

In [11]:
date_cols = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

for col in date_cols:
    orders[col] = pd.to_datetime(orders[col])

### ✅ Output Interpretation

The selected columns have been successfully converted to datetime format.

This transformation allows for efficient time-based operations such as filtering by date, calculating time differences, and analyzing order processing and delivery performance in subsequent steps.

### 🔗 Merging Orders and Customers Datasets

In this step, we perform a left join between the orders and customers datasets using the common key `customer_id`. This allows us to combine order-related information with corresponding customer details into a single DataFrame.

A left join ensures that all records from the orders dataset are retained, while matching customer information is appended wherever available.

In [12]:
df = orders.merge(customers, on="customer_id", how="left")
df.shape

(99441, 12)

### ✅ Output Interpretation

The merged DataFrame contains 99,441 rows and 12 columns.

This indicates that all order records have been preserved, and additional customer-related attributes have been successfully integrated. The increase in the number of columns reflects the inclusion of customer information.

This combined dataset will be used for deeper analysis, such as customer behavior, geographic trends, and order patterns.

### 🔗 Integrating Order Items Data

In this step, we merge the existing DataFrame with the order items dataset using the common key `order_id`. This allows us to incorporate product-level details such as item quantity, price, and seller information for each order.

Since a single order can contain multiple items, this merge expands the dataset to a more granular level, capturing each item within an order.

In [13]:
df = df.merge(items, on="order_id", how="left")
df.shape

(113425, 18)

### ✅ Output Interpretation

The resulting DataFrame now contains 113,425 rows and 18 columns.

The increase in the number of rows indicates that some orders consist of multiple items, leading to a one-to-many relationship between orders and order items. Each row now represents an individual item within an order rather than just the order itself.

This enriched dataset enables detailed analysis at the product and transaction level, such as revenue contribution, item-wise sales, and seller performance.

### 🔗 Incorporating Payment Information

In this step, we merge the current DataFrame with the payments dataset using the common key `order_id`. This integration allows us to include payment-related details such as payment type, installments, and transaction value for each order.

This step is essential for analyzing revenue patterns, payment preferences, and overall financial metrics.

In [14]:
df = df.merge(payments, on="order_id", how="left")
df.shape

(118434, 22)

### ✅ Output Interpretation

The DataFrame now contains 118,434 rows and 22 columns.

The increase in the number of rows suggests that some orders have multiple payment records (e.g., split payments or multiple payment methods), resulting in a one-to-many relationship between orders and payments.

With payment data integrated, the dataset is now better suited for financial analysis, including revenue calculation, payment behavior insights, and transaction-level evaluation.

### 🔗 Adding Customer Review Data

In this step, we merge the current DataFrame with the reviews dataset using the common key `order_id`. This allows us to incorporate customer feedback, including review scores and comments, for each order.

Integrating review data is crucial for analyzing customer satisfaction, product/service quality, and overall user experience.

In [15]:
df = df.merge(reviews, on="order_id", how="left")
df.shape

(119143, 28)

### ✅ Output Interpretation

The DataFrame now contains 119,143 rows and 28 columns.

The slight increase in the number of rows indicates that some orders may have multiple review entries or associated feedback records. This reflects a one-to-many relationship between orders and reviews.

With review data included, the dataset becomes more comprehensive, enabling sentiment analysis, satisfaction scoring, and deeper insights into customer experience.

### 🔗 Integrating Product Information

In this step, we merge the existing DataFrame with the products dataset using the common key `product_id`. This allows us to enrich the dataset with product-level attributes such as category, weight, dimensions, and other specifications.

Adding product data enables deeper analysis of sales performance across different product categories and characteristics.

In [16]:
df = df.merge(products, on="product_id", how="left")
df.shape

(119143, 36)

In [ ]:
### ✅ Output Interpretation

The DataFrame now contains 119,143 rows and 36 columns.

The number of rows remains unchanged, indicating a one-to-one relationship between order items and product details. Each item is now successfully associated with its corresponding product information.

This enriched dataset supports advanced analysis such as category-wise sales trends, product performance evaluation, and inventory-related insights.

### 🌐 Adding Product Category Translations

In this step, we merge the dataset with the product category translation table using the `product_category_name` column. This allows us to convert product category names from Portuguese to English.

This transformation improves readability and makes the dataset more accessible for analysis and presentation, especially for a broader audience.

In [17]:
df = df.merge(
    translation,
    on="product_category_name",
    how="left"
)
df.shape

(119143, 37)

### ✅ Output Interpretation

The DataFrame now contains 119,143 rows and 37 columns.

The number of rows remains unchanged, indicating that the merge was successfully performed without altering the dataset structure. A new column with translated category names has been added.

This enhancement makes it easier to interpret product categories and perform category-level analysis in a more understandable format.

### 🔍 Inspecting Dataset Structure and Data Types

In this step, we use the `.info()` function to get a comprehensive overview of the DataFrame. This includes information about the number of entries, column names, data types, and non-null counts.

This step is crucial for identifying missing values, understanding data types, and planning data cleaning and preprocessing strategies.

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119143 entries, 0 to 119142
Data columns (total 37 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       119143 non-null  object        
 1   customer_id                    119143 non-null  object        
 2   order_status                   119143 non-null  object        
 3   order_purchase_timestamp       119143 non-null  datetime64[ns]
 4   order_approved_at              118966 non-null  datetime64[ns]
 5   order_delivered_carrier_date   117057 non-null  datetime64[ns]
 6   order_delivered_customer_date  115722 non-null  datetime64[ns]
 7   order_estimated_delivery_date  119143 non-null  datetime64[ns]
 8   customer_unique_id             119143 non-null  object        
 9   customer_zip_code_prefix       119143 non-null  int64         
 10  customer_city                  119143 non-null  object        
 11  

### ✅ Output Interpretation

The dataset contains 119,143 entries and 37 columns, with a mix of data types including datetime, numerical, and categorical variables.

Key observations:
- Several columns contain missing values, particularly in delivery dates, product details, and review-related fields.
- Review comment fields have a high proportion of missing data, indicating that many customers did not leave textual feedback.
- Date columns are correctly formatted as datetime, enabling time-based analysis.
- The dataset includes a diverse range of features covering orders, customers, payments, products, and reviews.

This overview highlights the need for handling missing values and confirms that the dataset is well-structured for further exploratory data analysis and feature engineering.

### ⏱️ Feature Engineering: Delivery and Approval Metrics

In this step, we create new features to capture key time-based performance metrics in the order lifecycle.

- **delivery_days**: Calculates the total number of days taken to deliver an order from the purchase date.
- **approval_delay**: Measures the time taken for order approval after purchase.
- **late_delivery**: A binary indicator that flags whether an order was delivered later than the estimated delivery date.

These engineered features are essential for analyzing operational efficiency, delivery performance, and customer satisfaction.

In [19]:
df["delivery_days"] = (
    df["order_delivered_customer_date"] -
    df["order_purchase_timestamp"]
).dt.days

df["approval_delay"] = (
    df["order_approved_at"] -
    df["order_purchase_timestamp"]
).dt.days

df["late_delivery"] = (
    df["order_delivered_customer_date"] >
    df["order_estimated_delivery_date"]
).astype(int)

### ✅ Output Interpretation

New columns have been successfully added to the dataset:

- **delivery_days** and **approval_delay** enable quantitative analysis of time delays in the order process.
- **late_delivery** provides a clear indicator for identifying delayed deliveries, which can be used for performance tracking and predictive modeling.

These features enhance the dataset by enabling deeper insights into logistics efficiency and service quality.

### 💰 Feature Engineering: Financial Metrics

In this step, we create additional features to capture key financial aspects of each transaction.

- **total_order_value**: Represents the total cost of an order, including product price and freight charges.
- **freight_ratio**: Measures the proportion of freight cost relative to the product price, helping assess shipping cost impact.
- **value_per_installment**: Calculates the average payment amount per installment for orders paid in multiple parts.

These features are important for analyzing revenue structure, cost distribution, and customer payment behavior.

In [20]:
df["total_order_value"] = df["price"] + df["freight_value"]

df["freight_ratio"] = df["freight_value"] / df["price"]

df["value_per_installment"] = (
    df["payment_value"] / df["payment_installments"]
)

### ✅ Output Interpretation

The new financial features have been successfully added to the dataset:

- **total_order_value** provides a complete view of transaction value.
- **freight_ratio** helps identify orders with disproportionately high shipping costs.
- **value_per_installment** offers insight into customer payment patterns, especially for installment-based purchases.

These enhancements enable more detailed financial analysis and support advanced business insights related to pricing, logistics, and payment trends.

### ⚠️ Feature Engineering: Return Risk Indicator

In this step, we create a new feature called **return_risk** to identify potentially problematic orders.

This feature is based on two key conditions:
- Orders with **low review scores (≤ 2)**, indicating customer dissatisfaction
- Orders with **late delivery**, reflecting poor service experience

By combining these conditions, we create a binary indicator to flag orders that are more likely to result in returns or negative outcomes.

In [21]:
df["return_risk"] = (
    (df["review_score"] <= 2) |
    (df["late_delivery"] == 1)
).astype(int)

### ✅ Output Interpretation

The **return_risk** column has been successfully created as a binary variable:

- **1** → High risk (low review score or late delivery)
- **0** → Low risk

This feature is particularly valuable for predictive modeling and risk analysis, as it helps identify orders that may lead to returns, complaints, or customer dissatisfaction.

### 📅 Feature Engineering: Time-Based Purchase Patterns

In this step, we extract time-based features from the order purchase timestamp to analyze customer purchasing behavior.

- **purchase_month**: Captures the month in which the order was placed.
- **purchase_day**: Represents the day of the week when the purchase occurred (0 = Monday, 6 = Sunday).

These features help identify seasonal trends and weekly purchasing patterns.

In [22]:
df["purchase_month"] = df["order_purchase_timestamp"].dt.month
df["purchase_day"] = df["order_purchase_timestamp"].dt.dayofweek

### ✅ Output Interpretation

The new time-based features have been successfully added:

- **purchase_month** enables analysis of monthly or seasonal sales trends.
- **purchase_day** allows examination of customer activity across different days of the week.

These features are useful for understanding demand patterns and can support business decisions related to marketing, inventory planning, and promotions.

### 💾 Exporting the Final Processed Dataset

In this step, we export the fully processed and feature-engineered DataFrame to a CSV file named `master_olist.csv`.

This file serves as a consolidated dataset that combines all relevant information from multiple tables along with newly created features. It is optimized for further analysis, visualization, and modeling.

In [23]:
df.to_csv("master_olist.csv", index=False)

### ✅ Output Interpretation

The dataset has been successfully saved as `master_olist.csv` in the working directory.

This final dataset can now be used for:
- Data visualization (e.g., Power BI, Tableau)
- Advanced analytics and reporting
- Machine learning and predictive modeling

Exporting the data ensures reproducibility and provides a clean, ready-to-use dataset for downstream tasks.

### 📊 Descriptive Statistics of Key Features

In this step, we generate summary statistics for selected key variables: **delivery_days**, **total_order_value**, and **return_risk**.

This provides insights into the central tendency, variability, and distribution of important operational and financial metrics, helping us understand overall patterns in the dataset.

In [24]:
df[[
"delivery_days",
"total_order_value",
"return_risk"
]].describe()

,delivery_days,total_order_value,return_risk
count,115722.000000,118310.000000,119143.000000
mean,12.022589,140.678990,0.199835
std,9.454922,191.239906,0.399878
min,0.000000,6.080000,0.000000
25%,6.000000,55.240000,0.000000
50%,10.000000,91.990000,0.000000
75%,15.000000,157.615000,0.000000
max,209.000000,6929.310000,1.000000


### ✅ Output Interpretation

The summary statistics reveal several important insights:

- **Delivery Days**:
  - Average delivery time is approximately 12 days.
  - There is high variability, with some deliveries taking up to 209 days, indicating potential outliers or extreme delays.

- **Total Order Value**:
  - The average order value is around 140.68.
  - The distribution is right-skewed, with a maximum value significantly higher than the median, suggesting the presence of high-value orders.

- **Return Risk**:
  - The mean value of ~0.20 indicates that approximately 20% of orders are classified as high risk.
  - Since this is a binary variable, the standard deviation reflects the proportion of risky vs non-risky orders.

These insights provide a foundational understanding of delivery performance, revenue distribution, and potential risk areas in the dataset.

### 🗄️ Storing the Final Dataset in PostgreSQL

In this step, we load the fully processed and feature-engineered dataset into a PostgreSQL database table named `olist_master`.

This allows us to perform advanced SQL queries, build dashboards, and integrate the dataset with other analytical tools. Storing the data in a relational database enhances scalability and enables efficient data retrieval.

In [25]:
df.to_sql(
    "olist_master",
    engine,
    if_exists="replace",
    index=False
)

573

### ✅ Output Interpretation

The output indicates the number of rows successfully inserted into the PostgreSQL table.

This confirms that the dataset has been transferred from the Pandas DataFrame to the database without errors. The `olist_master` table is now ready for SQL-based analysis, reporting, and integration with visualization tools such as Power BI or Tableau.

### 🔍 Verifying Data Storage in PostgreSQL

In this step, we execute a SQL query to count the total number of records in the `olist_master` table stored in PostgreSQL.

This serves as a validation step to ensure that the dataset has been successfully loaded into the database without any data loss.

In [26]:
import pandas as pd
pd.read_sql("SELECT COUNT(*) FROM olist_master", engine)

,count
0,119143


### ✅ Output Interpretation

The query returns a total of 119,143 records in the `olist_master` table.

This matches the number of rows in the original DataFrame, confirming that the data has been accurately and completely stored in the database.

The dataset is now fully ready for SQL-based analysis, querying, and integration with visualization tools.

### 📊 Top 10 Revenue-Generating Product Categories

We run the following query to identify the top 10 product categories contributing the highest revenue.

This helps in understanding which product segments drive the majority of sales and where the business should focus its strategic efforts.

In [27]:
query = """
SELECT product_category_name_english,
       ROUND(SUM(price)::numeric, 2) AS revenue
FROM olist_master
GROUP BY product_category_name_english
ORDER BY revenue DESC
LIMIT 10;
"""

pd.read_sql(query, engine)

,product_category_name_english,revenue
0,health_beauty,1301947.97
1,watches_gifts,1254322.95
2,bed_bath_table,1107249.09
3,sports_leisure,1029603.88
4,computers_accessories,950053.69
5,furniture_decor,772096.17
6,housewares,668880.94
7,cool_stuff,664637.13
8,auto,618395.50
9,garden_tools,519473.33


### 📈 Key Insights & Business Interpretation

#### 🔍 Observations:
- **Health & Beauty** is the highest revenue-generating category (~1.30M), followed closely by **Watches & Gifts**.
- Categories like **Bed, Bath & Table** and **Sports & Leisure** also contribute significantly, each exceeding 1M in revenue.
- Mid-tier categories such as **Computers & Accessories** and **Furniture & Decor** show strong but comparatively lower performance.
- Long-tail categories like **Auto** and **Garden Tools** generate relatively less revenue.

#### 💡 Business Insights:
- The platform’s revenue is **highly concentrated in lifestyle and personal-use categories**.
- Categories related to **daily use and gifting** show consistently high demand.
- Technology-related products contribute well but are not the top drivers.

#### 🎯 Strategic Recommendations:
- ✅ Increase inventory and marketing investment in top-performing categories (Health & Beauty, Watches & Gifts)
- ✅ Bundle or cross-sell products within high-performing segments to increase average order value
- ✅ Explore growth opportunities in mid-performing categories through targeted promotions
- ⚠️ Re-evaluate low-performing categories for profitability, pricing strategy, or customer demand

#### 📊 Conclusion:
A small number of product categories contribute a significant portion of total revenue, indicating a **Pareto-like distribution (80/20 rule)**. Focusing on these key categories can maximize overall business growth.

### 📅 Monthly Revenue Trend Analysis

In this step, we analyze how revenue changes over time by aggregating total sales on a monthly basis.

The `DATE_TRUNC('month', ...)` function is used to group transactions by month, allowing us to identify trends, seasonality, and growth patterns in the business.

In [29]:
query = """
SELECT 
    DATE_TRUNC('month', order_purchase_timestamp) AS month,
    ROUND(SUM(price)::numeric, 2) AS revenue
FROM olist_master
GROUP BY month
ORDER BY month;
"""

pd.read_sql(query, engine)

,month,revenue
0,2016-09-01,267.36
1,2016-10-01,51068.92
2,2016-12-01,10.90
3,2017-01-01,129895.32
4,2017-02-01,262013.86
5,2017-03-01,398117.44
6,2017-04-01,392595.36
7,2017-05-01,549226.84
8,2017-06-01,456867.47
9,2017-07-01,536906.96


### 📈 Monthly Revenue Trends – Insights & Business Interpretation

#### 🔍 Key Observations:
- Revenue shows **very low activity in 2016**, indicating the early stage of the platform.
- From **January 2017 onwards**, revenue starts increasing steadily, showing strong business growth.
- A significant spike is observed in **November 2017 (~1.05M)**, likely due to major sales events (e.g., Black Friday).
- Revenue remains consistently high throughout **2018**, mostly around ~900K–1M per month, indicating business maturity.
- A sharp drop is seen in **September 2018 and October 2018**, which may indicate:
  - Incomplete data
  - Data collection cutoff
  - Or business disruption

#### 📊 Trend Analysis:
- The business demonstrates a **clear upward growth trajectory from 2017 to 2018**
- Presence of **seasonality**, especially during year-end sales periods
- Revenue stabilizes after rapid growth, indicating a **transition from growth phase to stable phase**

#### 💡 Business Insights:
- The company successfully scaled operations during 2017
- Peak sales periods (especially November) are **critical revenue drivers**
- Consistent high revenue in 2018 suggests strong customer retention and demand

#### 🎯 Strategic Recommendations:
- ✅ Focus heavily on **festive/seasonal campaigns** (especially Q4)
- ✅ Optimize inventory and logistics for high-demand months like November
- ✅ Investigate the sudden drop in late 2018 (data quality or business issue)
- ✅ Use this trend data for **sales forecasting and capacity planning**

#### ⚠️ Data Quality Note:
- The unusually low/NaN values in the final months suggest **incomplete or truncated data**, which should be considered during analysis.

### 🌍 Revenue Distribution by Customer State

In this step, we analyze how revenue is distributed across different customer states.

This helps identify high-performing regions and understand geographical demand patterns, which is crucial for regional marketing, logistics planning, and business expansion strategies.

In [30]:
query = """
SELECT 
    customer_state,
    ROUND(SUM(price)::numeric, 2) AS revenue
FROM olist_master
GROUP BY customer_state
ORDER BY revenue DESC
LIMIT 10;
"""

pd.read_sql(query, engine)

,customer_state,revenue
0,SP,5477008.74
1,RJ,1921752.53
2,MG,1645847.27
3,RS,791716.84
4,PR,708794.22
5,BA,543243.99
6,SC,539896.20
7,DF,315122.29
8,GO,313198.27
9,ES,284771.30


### 📈 Key Insights & Business Interpretation

#### 🔍 Observations:
- **São Paulo (SP)** dominates revenue generation (~5.47M), contributing significantly more than any other state.
- **Rio de Janeiro (RJ)** and **Minas Gerais (MG)** follow as the next top contributors, but at a much lower scale.
- Southern states like **RS** and **PR** also show moderate performance.
- Other states contribute relatively smaller portions of total revenue.

#### 📊 Trend Analysis:
- Revenue is **highly concentrated in a few major states**, especially SP.
- There is a sharp drop from the top state (SP) to others, indicating **regional imbalance in sales**.

#### 💡 Business Insights:
- São Paulo acts as the **primary revenue hub**, likely due to higher population, urbanization, and purchasing power.
- Other states represent **growth opportunities** rather than saturated markets.
- Logistics and delivery infrastructure may be more optimized in top-performing regions.

#### 🎯 Strategic Recommendations:
- ✅ Strengthen logistics and inventory management in **SP** to maintain performance
- ✅ Invest in targeted marketing campaigns in **mid-performing states (RJ, MG)** to boost revenue
- ✅ Explore expansion strategies in underperforming regions to increase market penetration
- ✅ Customize regional offers based on demand patterns

#### 📊 Conclusion:
The business shows a **strong geographic concentration**, with a few states driving the majority of revenue. Diversifying regional performance can unlock additional growth potential.

### ⚠️ Return Risk Analysis by Customer State

In this step, we calculate the return risk percentage across different customer states.

The `return_risk` flag (derived earlier) identifies potentially problematic orders based on:
- Low review scores (≤ 2)
- Late deliveries

We use conditional aggregation to compute the percentage of high-risk orders in each state.

In [31]:
query = """
SELECT
    customer_state,
    COUNT(*) FILTER (WHERE return_risk = 1) * 100.0 / COUNT(*) AS return_rate_pct
FROM olist_master
GROUP BY customer_state
ORDER BY return_rate_pct DESC
LIMIT 10;
"""

pd.read_sql(query, engine)

,customer_state,return_rate_pct
0,AL,34.698276
1,MA,32.359813
2,SE,27.791563
3,RR,26.923077
4,RJ,26.768914
5,BA,26.350526
6,CE,26.261981
7,PI,26.041667
8,PA,25.686448
9,ES,22.584746


### 📈 Key Insights & Business Interpretation

#### 🔍 Observations:
- **Alagoas (AL)** shows the highest return risk (~34.7%), followed by **Maranhão (MA)** and **Sergipe (SE)**.
- Several states exhibit return rates above **25%**, indicating a significant proportion of problematic orders.
- Even major markets like **Rio de Janeiro (RJ)** show relatively high return risk (~26.7%).

#### 📊 Trend Analysis:
- Return risk is **not evenly distributed geographically**
- Certain regions consistently show **higher operational or customer satisfaction issues**

#### 💡 Business Insights:
- High return risk may be driven by:
  - Poor delivery performance
  - Product quality issues
  - Customer expectation mismatch
- States with high return rates may also suffer from **logistics inefficiencies or last-mile delivery challenges**

#### 🎯 Strategic Recommendations:
- ⚠️ Investigate logistics partners and delivery timelines in high-risk states (AL, MA, SE)
- ✅ Improve customer experience through better product descriptions and quality checks
- ✅ Introduce stricter seller quality controls in high-risk regions
- ✅ Implement proactive customer support for delayed deliveries

#### 📊 Conclusion:
Return risk varies significantly across regions, highlighting **operational inefficiencies and customer experience gaps**. Addressing these issues can reduce losses and improve customer satisfaction.

### 💸 Revenue Loss Analysis by Product Category

In this step, we analyze revenue loss due to high return-risk orders across product categories.

We calculate:
- Returned revenue (from high-risk orders)
- Total revenue
- Percentage revenue loss (`loss_pct`)

This helps identify which categories are most impacted by returns and operational inefficiencies.

In [33]:
query = """
SELECT
    product_category_name_english,
    ROUND(SUM(price) FILTER (WHERE return_risk=1)::numeric, 2) AS returned_revenue,
    ROUND(SUM(price)::numeric, 2) AS total_revenue,
    ROUND(
        (
            SUM(price) FILTER (WHERE return_risk=1)::numeric
            / SUM(price)::numeric
        ) * 100, 2
    ) AS loss_pct
FROM olist_master
GROUP BY product_category_name_english
ORDER BY returned_revenue DESC
LIMIT 10;
"""

pd.read_sql(query, engine)

,product_category_name_english,returned_revenue,total_revenue,loss_pct
0,cds_dvds_musicals,NaN,730.00,NaN
1,bed_bath_table,250571.88,1107249.09,22.63
2,watches_gifts,238048.43,1254322.95,18.98
3,health_beauty,222058.40,1301947.97,17.06
4,computers_accessories,217378.03,950053.69,22.88
5,sports_leisure,202666.94,1029603.88,19.68
6,furniture_decor,179258.75,772096.17,23.22
7,auto,157693.36,618395.50,25.50
8,housewares,127481.19,668880.94,19.06
9,cool_stuff,126021.81,664637.13,18.96


### 📉 Revenue Loss Analysis – Insights & Business Interpretation

#### 🔍 Observations:
- Categories like **Bed, Bath & Table**, **Watches & Gifts**, and **Health & Beauty** have the highest absolute revenue loss due to returns.
- **Auto (~25.5%)** and **Furniture & Decor (~23.2%)** show the highest **loss percentage**, indicating a larger proportion of problematic orders.
- **Computers & Accessories (~22.9%)** and **Bed, Bath & Table (~22.6%)** also exhibit significant loss ratios.
- Some categories like **CDs, DVDs & Musicals** show `NaN` values, likely due to:
  - No return-risk orders
  - Or very low transaction volume

#### 📊 Trend Analysis:
- High-revenue categories are also experiencing **high return-related losses**
- Certain categories have **disproportionately high loss percentages**, indicating deeper operational or product issues

#### 💡 Business Insights:
- Categories with high loss % (Auto, Furniture) may suffer from:
  - Product quality issues
  - Delivery damage (especially for bulky items)
  - Customer expectation mismatch
- High-volume categories (Health & Beauty, Watches) generate more absolute loss due to scale

#### 🎯 Strategic Recommendations:
- ⚠️ Investigate **Auto & Furniture categories** for quality control and logistics handling
- ✅ Improve packaging and delivery processes for fragile/bulky items
- ✅ Optimize product descriptions and images to reduce expectation gaps
- ✅ Implement stricter seller quality checks in high-loss categories
- ✅ Consider return policies or restocking fees for high-risk segments

#### 📊 Conclusion:
Revenue loss due to return-risk orders is **significant and category-dependent**. Addressing high-loss categories can directly improve profitability and operational efficiency.

### 🚚 Impact of Delivery Speed on Return Risk

In this step, we categorize orders based on delivery speed using `delivery_days`:

- Fast (≤ 3 days)
- Moderate (4–7 days)
- Slow (8–14 days)
- Very Slow (> 14 days)

We then analyze how return risk varies across these delivery speed categories to understand the impact of logistics performance on customer satisfaction and returns.

In [34]:
query = """
SELECT
    CASE
        WHEN delivery_days <= 3 THEN 'Fast'
        WHEN delivery_days <= 7 THEN 'Moderate'
        WHEN delivery_days <= 14 THEN 'Slow'
        ELSE 'Very Slow'
    END AS delivery_speed,

    COUNT(*) AS orders,
    ROUND(
        COUNT(*) FILTER (WHERE return_risk=1) * 100.0
        / COUNT(*), 2
    ) AS return_rate_pct

FROM olist_master
WHERE delivery_days IS NOT NULL
GROUP BY delivery_speed
ORDER BY return_rate_pct DESC;
"""

pd.read_sql(query, engine)

,delivery_speed,orders,return_rate_pct
0,Very Slow,31282,36.86
1,Slow,43943,12.54
2,Moderate,30139,10.72
3,Fast,10358,9.36


### 📈 Key Insights & Business Interpretation

#### 🔍 Observations:
- **Very Slow deliveries (>14 days)** have the highest return rate (~36.86%), which is significantly higher than all other categories.
- Return rate drops sharply as delivery speed improves:
  - Slow: ~12.54%
  - Moderate: ~10.72%
  - Fast: ~9.36%
- The number of orders is also highest in the **Slow and Very Slow categories**, amplifying the overall business impact.

#### 📊 Trend Analysis:
- There is a **strong positive correlation between delivery time and return risk**
- As delivery delays increase, customer dissatisfaction and return probability increase drastically

#### 💡 Business Insights:
- Delivery speed is a **critical driver of customer satisfaction and retention**
- Late deliveries significantly contribute to negative reviews and return-risk classification
- Logistics inefficiency directly impacts revenue loss and brand perception

#### 🎯 Strategic Recommendations:
- ⚠️ Urgently optimize logistics for orders taking >14 days (Very Slow category)
- ✅ Improve last-mile delivery and warehouse efficiency
- ✅ Set realistic delivery expectations to avoid customer dissatisfaction
- ✅ Prioritize faster shipping options for high-value or high-risk orders
- ✅ Monitor delivery KPIs as a core business performance metric

#### 📊 Conclusion:
Delivery performance has a **direct and substantial impact on return risk**. Reducing delivery time can significantly lower returns and improve overall customer experience.

## 🏁 Final Conclusion & Business Summary

This analysis of the Brazilian E-Commerce dataset provides a comprehensive understanding of sales performance, customer behavior, operational efficiency, and return risk.

---

### 📊 Key Takeaways:

- 📈 **Strong Business Growth**: The platform experienced rapid growth from 2017 to 2018, followed by a phase of stable high revenue.
- 🛍️ **Top Revenue Drivers**: Categories like *Health & Beauty*, *Watches & Gifts*, and *Bed, Bath & Table* contribute the most to total revenue.
- 🌍 **Geographical Concentration**: A significant portion of revenue is concentrated in a few key states, especially São Paulo.
- ⚠️ **Return Risk Factors**:
  - High return risk is linked to **low customer satisfaction (low review scores)** and **late deliveries**
  - Certain states and categories show disproportionately high return rates
- 🚚 **Logistics Impact**: Delivery delays have a **direct and significant impact on return risk**, with very slow deliveries showing nearly 4x higher return rates.
- 💸 **Revenue Leakage**: High-performing categories also experience substantial revenue loss due to returns, highlighting areas for optimization.

---

### 💡 Business Recommendations:

- ✅ Optimize **logistics and delivery performance** to reduce return risk
- ✅ Focus on **high-revenue categories** for growth and cross-selling strategies
- ✅ Improve **product quality, packaging, and descriptions** to minimize returns
- ✅ Expand strategically in **underperforming regions** to diversify revenue
- ✅ Monitor key KPIs like **return rate, delivery time, and customer satisfaction**

---

### 🎯 Final Thoughts:

This project demonstrates how data-driven analysis can uncover critical business insights and support strategic decision-making.

By combining SQL, Python, and business understanding, we transformed raw transactional data into actionable insights that can drive revenue growth, reduce losses, and enhance customer experience.

---

### 🚀 Tools & Technologies Used:
- Python (Pandas, SQLAlchemy)
- PostgreSQL
- Jupyter Notebook
- Power BI (for dashboard visualization)

---

> 📌 *This project reflects end-to-end data analytics workflow — from data cleaning and feature engineering to advanced SQL analysis and business insight generation.*